mvc/views/tag-helpers/authoring.rst

```diff
@@ -1,8 +1,394 @@
+- The last line sets the completed content for our minimally functional tag helper.
+- The following line shows the syntax for adding attributes:
+
+.. literalinclude:: authoring/sample/AuthoringTagHelpers/src/AuthoringTagHelpers/TagHelpers/EmailTagHelperMailTo.cs
+   :lines: 14-21
+   :dedent: 8
+   :emphasize-lines: 6
+   
+That approach works for the attribute "href" as long as it doesn't currently exist in the attributes collection. You can also use the ``output.Attributes.Add`` method to add a tag helper attribute to the end of the collection of tag attributes.
+
+3. Update the markup in the *Views/Home/Contact.cshtml* file with these changes:
+
+.. literalinclude::  authoring/sample/AuthoringTagHelpers/src/AuthoringTagHelpers/Views/Home/ContactCopy.cshtml
+   :language: aspx-cs
+   :emphasize-lines: 15-16 
+   :lines: 1-17
+ 
+4. Run the app and verify that it generates the correct links.
+
+**Note:** If you were to write the email tag self-closing (``<email mail-to="Rick" />``), the final output would also be self-closing. To enable the ability to write the tag with only a start tag ( <email mail-to="Rick"> ) you must decorate the class with the following:
+
+.. literalinclude:: authoring/sample/AuthoringTagHelpers/src/AuthoringTagHelpers/TagHelpers/EmailTagHelperMailVoid.cs
+   :lines: 6
+   :dedent: 4 
+   
+In our example, the output would be ``<a href="mailto:Rick@contoso.com" />``. Self-closing anchor tags are not valid HTML, so you wouldn't want to create one, but you might want to create a tag helper that is self-closing. The ASP.NET 5 runtime sets the type of the ``TagMode`` property after reading a tag.
+
+.. In this section we will update the ``EmailTagHelper`` so that it gets the target ``mail-to`` from the content. Need to revert the contact.cshtml file back.
+
+An asynchronous email helper
+_____________________________________
+
+In this section we'll write an asynchronous email helper. 
+
+#. Replace the ``EmailTagHelper`` class with the following code:
+
+.. literalinclude:: authoring/sample/AuthoringTagHelpers/src/AuthoringTagHelpers/TagHelpers/EmailTagHelper.cs
+   :lines: 6-17
+   :dedent: 4
+
+**Notes:**
+
+- This version uses the asynchronous ``ProcessAsync`` method. The asynchronous ``GetChildContentAsync`` returns a ``Task`` containing the ``TagHelperContent``. 
+- We use the ``context`` parameter to get contents of the HTML element.
+
+2. Make the following change to the *Views/Home/Contact.cshtml* file so the tag helper can get the target email.
+
+.. literalinclude::  authoring/sample/AuthoringTagHelpers/src/AuthoringTagHelpers/Views/Home/Contact.cshtml
+   :language: aspx-cs
+   :emphasize-lines: 15-16 
+   :lines: 1-17
+
+3. Run the app and verify that it generates a valid email link.
```

Comment:
`Run the app and verify that it generates a valid email link.` => `Run the app and verify that it generates valid email links.`